node-neigh: Fix unit test flake

brb opened the pull request:
See commit msgs.

Fix #16040 
Fix #16075

brb committed:
node-neigh: Set lastPing if arping was successful

We don't return early if arping was skipped. This can happen when
insertNeighbor() is invoked by the non-refresh path and nexthop is not
new.

Make sure that lastPing is updated only if arping was sent and it was
successful (if hwAddr != nil condition).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Wait instead of sleeping in unit tests

We can inspect the neighLastPingByNextHop map to check when
insertNeighbor() or deleteNeighbor() was called.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

kkourt commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -1026,18 +1027,41 @@ func (s *linuxPrivilegedIPv4OnlyTestSuite) TestArpPingHandling(c *check.C) {
+	now := time.Now()
 	err = linuxNodeHandler.NodeAdd(nodev1)
 	c.Assert(err, check.IsNil)
+	// insertNeighbor is invoked async
```
The same comment appears a few lines below. Was this maybe forgotten from an earlier version?

brb replied:
Removed redundant comments.

kkourt commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -1026,18 +1027,41 @@ func (s *linuxPrivilegedIPv4OnlyTestSuite) TestArpPingHandling(c *check.C) {
 	c.Assert(err, check.IsNil)
 
+	wait := func(now time.Time, nodeID nodeTypes.Identity, isDelete bool) {
+		err := testutils.WaitUntil(func() bool {
```
From my reading of the code `now` is a lower bound for when the operation happened.

However, it is used only for insertion and not deletion (since for the latter, we do not have a deletion timestamp to check).
Given this, and the overall structure of the function, I wonder if it would be better to have two versions of this function one for deletes and one for inserts.

I think this would also improve readability on the caller side (at the cost of some code duplication).

brb replied:
> From my reading of the code now is a lower bound for when the operation happened.

Did s/now/before/. Hope it's more clear.

> However, it is used only for insertion and not deletion

Made `before` as `*time.Time`, so that nil can be passed for the wait for the deletion.

> Given this, and the overall structure of the function, I wonder if it would be better to have two versions of this function one for deletes and one for inserts.

I want this waiting routine to be concise to avoid polluting the test case. Also, to follow DRY principle.

kkourt replied:
OK, sounds good.

brb committed:
node-neigh: Skip neigh update if arping is stale

It's possible that in the case of multiple concurrent insertNeighbor()
executions the oldest (or older) goroutine will overwrite the latest
arping result due to the fine-grained locking.

To fix this, avoid updating neigh entry if we detect that prev last ping
timestamp is after our arping timestamp.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

kkourt reviewed (approved):
Last patch looks good to me as well.

Out of curiosity, was this triggered because of the small refresh period:
```
option.Config.ARPPingRefreshPeriod = time.Duration(1 * time.Nanosecond)
```
